doc/DesignConsiderations.md

```diff
@@ -0,0 +1,120 @@
+
+  * Heap memory (if a dynamic memory allocator is present)
+  * Zero-initialized read-write data (the `.bss` section)
+  * Non-zero-initialized read-write data (the `.data` section)
+  * Stack memory
+
+`libtock_core` developers should consider the space usage of their code, and
+where the space usage is (non-volatile memory or RAM). Important note: the
+`.data` section consumes space in both non-volatile memory and RAM!
+
+Most apps will not use all functionality in `libtock_core`. The size impact of
+using `libtock_core` in an app should be commensurate with the `libtock_core`
+functionality that app uses. If done correctly, this will allow `libtock_core`'s
+users to build Tock systems with multiple small, single-purpose apps without
+excessive code duplication.
+
+## Testability
+
+Programmatic tests are important to verify that new functionality works
+correctly and to keep existing functionality working through nontrivial
+refactorings. Different types of tests offer different capabilities:
+
+  * **Host-based unit tests** run on a non-Tock host OS (such as Linux). They
+    can only be used with portable `libtock_core` code, as they code they test
+    must run on both Tock and the host OS. These tests must either test code
+    that does not directly use system calls, or must direct the system calls to
+    a fake Tock kernel. These tests can be fast (compile and run in seconds),
+    and can easily simulate a variety of failure modes (e.g. kernel errors) that
+    may be difficult to generate with a real Tock kernel.
+  * **Emulation tests** run on an emulated Tock system. Currently, emulation
+    tests run in QEMU on an emulated HiFive1 Rev B board. These test
+    `libtock_core`'s interaction with a real Tock kernel, albeit with limited
+    hardware access.
+  * **Hardware tests** run on physical hardware with a real Tock kernel. These
+    are end-to-end tests that test not only `libtock_core`'s functionality but
+    also the Tock kernel's interaction with the real hardware. However,
+    automating these tests is difficult, and currently we do not have a way to
+    run hardware tests in CI.
+
+If `libtock_core` is functional on real hardware with a real kernel, it should
+be able to support emulation tests and hardware tests. However, host-based unit
+tests require more work. In its intended use case, `libtock_core` is not
+supposed to be portable, supporting only the Tock kernel. Hoever, host-based
+unit tests are very valuable, both for rapid development and for testing error
+handling code. To support host-based tests, `libtock_core` must be built out of
+portable pieces that can be unit tested, even though the library as a whole is
+not portable.
+
+Testing pieces of `libtock_core` against a fake kernel can be done using
+dependency injection, but most dependency injection techniques have considerable
+code size and RAM usage costs. `libtock_core` needs a dependency injection
+technique that has minimal size impact.
```

Comment:
Do we need this?

Reply:
The only practical way that I am aware of to test system call error handling logic is through dependency injection. Do you know of another way?

Reply:
I meant should we say "**needs** a dependency injection..."

Reply:
I changed the sentence to the following:

`libtock_core` needs either a dependency injection technique that has minimal size impact or an alternative mechanism for testing `libtock_core`'s system call error handling logic.